src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -267,11 +268,90 @@ public class EpochManager : IEpochManager {
 				if (!_writer.Write(rec, out pos))
 					throw new Exception($"Second write try failed at {epoch.EpochPosition}.");
 			}
+			WriteEpochInformationWithRetry(epoch);
 			_writer.Flush();
+			_bus.Publish(new ReplicationTrackingMessage.WriterCheckpointFlushed());
 			_bus.Publish(new SystemMessage.EpochWritten(epoch));
 			return epoch;
 		}
 
+		void WriteEpochInformationWithRetry(EpochRecord epoch) {
+			if (!TryGetExpectedVersionForEpochInformation(epoch, out var expectedVersion))
+				expectedVersion = ExpectedVersion.NoStream;
+
+			var originalLogPosition = _writer.Checkpoint.ReadNonFlushed();
+
+			var epochInformation = LogRecord.Prepare(
+					logPosition: originalLogPosition,
+					correlationId: Guid.NewGuid(),
+					eventId: Guid.NewGuid(),
+					transactionPos: originalLogPosition,
+					transactionOffset: 0,
+					eventStreamId: SystemStreams.EpochInformationStream,
+					expectedVersion: expectedVersion,
+					flags: PrepareFlags.SingleWrite | PrepareFlags.IsCommitted | PrepareFlags.IsJson,
+					eventType: SystemEventTypes.EpochInformation,
+					data: epoch.AsSerialized(),
+					metadata: Empty.ByteArray);
+
+			if (_writer.Write(epochInformation, out var retryLogPosition))
+				return;
+
+			epochInformation = new PrepareLogRecord(
+				retryLogPosition,
+				epochInformation.CorrelationId,
+				epochInformation.EventId,
+				retryLogPosition,
+				epochInformation.TransactionOffset,
+				epochInformation.EventStreamId,
+				epochInformation.ExpectedVersion,
+				epochInformation.TimeStamp,
+				epochInformation.Flags,
+				epochInformation.EventType,
+				epochInformation.Data,
+				epochInformation.Metadata);
+
+			if (_writer.Write(epochInformation, out _))
+				return;
+
+			throw new Exception(
+				string.Format("Second write try failed when first writing $epoch-information at {0}, then at {1}.",
+					originalLogPosition,
+					retryLogPosition));
+		}
+
+		// we have just written epoch. about to write the $epoch-information for it.
+		// this decides what expected version the $epoch-information should have.
+		// it looks up the previous epoch, finds that epoch's previous information
+		// (which immediately follows it) and gets its event number.
+		bool TryGetExpectedVersionForEpochInformation(EpochRecord epoch, out long expectedVersion) {
+			expectedVersion = default;
```

Comment:
@timothycoleman if epoch.PrevEpochPosition is -1, we can directly return false here otherwise it will reposition the reader to -1 (i'm not sure if the reader's behaviour is well defined in this case)